File: framework/languages/node/e2e/ResilienceLifecycle/Client/Scenarios/rl-a1-provider-restart-scenario.ts
// RL-A1: 같은 endpoint에서 server를 재시작한다 시나리오를 검증한다.
import type { ClientOptions } from '../Support/client-options';
import { postJson } from '../../../http-client';
import { startProvider } from '../Support/managed-provider';
import {
  profileReq,
  waitForProviderTraffic,
  waitPeerAbsent,
  waitPeerPresent,
  waitUntilDown
} from '../Support/resilience-helpers';
import { ensure } from '../Support/scenario-assert';
import type { ScenarioState } from '../Support/scenario-state';
import type { ProfileRes } from '../../Shared/messages';

export async function runRlA1(options: ClientOptions, state: ScenarioState): Promise<void> {
  if (state.providerBProcess !== undefined) {
    await state.providerBProcess.stop();
  } else {
    await postJson(options.providerBUrl, '/shutdown');
  }
  await waitUntilDown(options.providerBUrl);
  await waitPeerAbsent(options.peerLocationUrl, 'api-b');
  state.providerBProcess = undefined;

  for (let i = 0; i < 12; i += 1) {
    const marker = `rl-a1-down-${i}`;
    const reply = await postJson<ProfileRes>(options.consumerUrl, '/profile/request', profileReq(marker));
    ensure(reply.providerRid === 'api-a', 'RL-A1 request during api-b restart did not use surviving provider.');
  }
  await postJson<string[]>(options.providerAUrl, '/evidence/wait', { contains: 'marker=rl-a1-down-' });

  const restarted = startProvider({
    providerMain: options.providerMain,
    logDir: options.logDir,
    redisEndpoint: options.redisEndpoint,
    redisKeyPrefix: options.redisKeyPrefix,
    name: 'api-b-restart',
    rid: 'api-b',
    httpUrl: options.providerBUrl,
    channelEndpoint: options.providerBChannelEndpoint,
    evidenceFileName: 'api-b-restart.evidence.log'
  });
  await restarted.waitReady();
  state.providerBProcess = restarted;
  await waitPeerPresent(options.peerLocationUrl, 'api-b');

  await waitForProviderTraffic(options.consumerUrl, 'rl-a1-restored', 'api-b');
  await postJson<string[]>(options.providerBUrl, '/evidence/wait', { contains: 'marker=rl-a1-restored-' });

  console.log('scenario RL-A1 passed');
}

File: framework/languages/dotnet/e2e/ResilienceLifecycle/Client/Scenarios/RlA1ProviderRestartScenario.cs
// Verifies RL-A1 Provider Restart behavior.
using ResilienceLifecycle.Client.Support;
using ResilienceLifecycle.Shared;
using Zlink.HttpClient;

namespace ResilienceLifecycle.Client.Scenarios;

// RL-A1 verifies recovery when the same provider endpoint is restarted.
internal static class RlA1ProviderRestartScenario
{
    public static async Task RunAsync(
        ZLinkHttpClient consumer,
        ZLinkHttpClient registry,
        ResilienceProcessManager processes,
        ZLinkHttpClient providerA,
        ZLinkHttpClient providerB)
    {
        // RL-A1 deliberately has one serving provider. api-a remains alive only
        // as harness infrastructure and is excluded from transport selection.
        await providerA.Post("/admin/weight/exclude").AsyncRaw();
        await providerA.Post("/admin/weight/wait").Body(new WeightWaitReq(0)).AsyncRaw();

        var oldRows = (await registry.Post("/topology/wait")
            .Body(new TopologyWaitReq("api-b", "Ready", 1))
            .Async<TopologyEntryRes[]>()).Body;
        var oldRoutingId = oldRows.Single().RoutingId;

        var acceptedMarker = $"rl-a1-accepted-{Guid.NewGuid():N}";
        var accepted = consumer.Post("/profile/request")
            .Body(new ProfileReq("slow", acceptedMarker))
            .Async<ProfileRes>();
        await providerB.Post("/evidence/wait")
            .Body(new EvidenceWaitReq([$"profile-start|rid=api-b|marker={acceptedMarker}"], []))
            .Async<string[]>();
        await processes.StopProviderBWithSigtermAsync();
        var acceptedReply = (await accepted).Body;
        ZlinkStreamAssert.Ensure(
            acceptedReply.ProviderRid == "api-b" && acceptedReply.Marker == acceptedMarker,
            "RL-A1 accepted request did not complete inside graceful shutdown.");

        await registry.Post("/topology/wait")
            .Body(new TopologyWaitReq("api-b", "Ready", 0))
            .Async<TopologyEntryRes[]>();

        var down = (await consumer.Post("/profile/request/attempt/1000")
            .Body(new ProfileReq("fast", "rl-a1-down"))
            .Async<ProfileAttemptRes>()).Body;
        // Spec 06 and the common E2E contract allow either NotFound for a
        // missing route or Unavailable for a registered route with no ready
        // target. Both are terminal here; the request must not be resent.
        ZlinkStreamAssert.Ensure(
            down is { Reply: null, ErrorKind: "NotFound" or "Unavailable" },
            $"RL-A1 down-window result was '{down.ErrorKind}', expected NotFound or Unavailable.");

        var connectionCount = (await consumer.Get("/connections").Async<string[]>()).Body.Length;
        var restarted = await processes.StartProviderBAsync();
        var newRows = (await registry.Post("/topology/wait")
            .Body(new TopologyWaitReq("api-b", "Ready", 1))
            .Async<TopologyEntryRes[]>()).Body;
        var newRow = newRows.Single();
        // Automatic discovery issues a new lifecycle RID on restart; the endpoint
        // is the one the provider was restarted on.
        ZlinkStreamAssert.Ensure(
            newRow.Endpoint == restarted.Endpoint && newRow.RoutingId != oldRoutingId,
            "RL-A1 restart did not publish the same endpoint under a new RID.");
        await consumer.Post("/connections/wait")
            .Body(new ConnectionWaitReq(
                ["kind=ConnectionReady", $"remote={restarted.Endpoint}"], connectionCount))
            .Async<string[]>();

        for (var i = 0; i < 20; i++)
        {
            var reply = (await consumer.Post("/profile/request")
                .Body(new ProfileReq("fast", $"rl-a1-restored-{i}"))
                .Async<ProfileRes>()).Body;
            ZlinkStreamAssert.Ensure(reply.ProviderRid == "api-b", "RL-A1 follow-up used an unexpected provider.");
        }

        await providerA.Post("/admin/weight/include").AsyncRaw();

        Console.WriteLine("scenario RL-A1 passed");
    }
}
